maze_transformer/generation/lattice_maze.py

```diff
@@ -306,17 +364,426 @@ def from_tokens(cls, tokens: list[str]) -> "LatticeMaze":
+
+        return "\n".join("".join(row) for row in ascii_grid)
+
+    @classmethod
+    def from_ascii(cls, ascii_str: str) -> "LatticeMaze":
+        lines: list[str] = ascii_str.strip().split("\n")
+        ascii_grid: Shaped[np.ndarray, "x y"] = np.array(
+            [list(line) for line in lines], dtype=str
+        )
+        pixel_grid: PixelGrid = np.zeros((*ascii_grid.shape, 3), dtype=np.uint8)
+
+        for ascii_char, pixel_color in ASCII_PIXEL_PAIRINGS.items():
+            pixel_grid[ascii_grid == ascii_char] = pixel_color
+
+        return cls.from_pixels(pixel_grid)
+
+
+@serializable_dataclass(frozen=True, kw_only=True)
+class TargetedLatticeMaze(LatticeMaze):
+    """A LatticeMaze with a start and end position"""
+
+    # this jank is so that SolvedMaze can inherit from this class without needing arguments for start_pos and end_pos
+    start_pos: Coord
+    end_pos: Coord
+
+    def __post_init__(self) -> None:
+        # make things numpy arrays (very jank to override frozen dataclass)
+        self.__dict__["start_pos"] = np.array(self.start_pos)
+        self.__dict__["end_pos"] = np.array(self.end_pos)
+        assert self.start_pos is not None
+        assert self.end_pos is not None
+        # check that start and end are in bounds
+        if (
+            self.start_pos[0] >= self.grid_shape[0]
+            or self.start_pos[1] >= self.grid_shape[1]
+        ):
+            raise ValueError(
+                f"start_pos {self.start_pos} is out of bounds for grid shape {self.grid_shape}"
+            )
+        if (
+            self.end_pos[0] >= self.grid_shape[0]
+            or self.end_pos[1] >= self.grid_shape[1]
+        ):
+            raise ValueError(
+                f"end_pos {self.end_pos} is out of bounds for grid shape {self.grid_shape}"
+            )
+
+    @classmethod
+    def from_lattice_maze(
+        cls,
+        lattice_maze: LatticeMaze,
+        start_pos: Coord,
+        end_pos: Coord,
+    ) -> "TargetedLatticeMaze":
+        return cls(
+            connection_list=lattice_maze.connection_list,
+            start_pos=start_pos,
+            end_pos=end_pos,
+        )
+
 
-class SolvedMaze(NamedTuple):
+@serializable_dataclass(frozen=True, kw_only=True)
+class SolvedMaze(TargetedLatticeMaze):
     """Stores a maze and a solution"""
 
-    maze: LatticeMaze
-    solution: list[CoordTup]
+    solution: CoordArray
+
+    def __init__(
+        self,
+        connection_list: ConnectionList,
+        solution: CoordArray,
+        generation_meta: dict | None = None,
+        lattice_dim: int = 2,
+    ) -> None:
+        super().__init__(
+            connection_list=connection_list,
+            generation_meta=generation_meta,
+            start_pos=np.array(solution[0]),
+            end_pos=np.array(solution[-1]),
+            lattice_dim=lattice_dim,
+        )
+        self.__dict__["solution"] = solution
```

Comment:
Also here, seems like this just doesn't need to be a frozen dataclass and this line should be `self.solution = solution`